var/spack/repos/builtin/packages/sim-model/package.py

```diff
@@ -193,6 +194,9 @@ def _install_src(self, prefix):
 
     def _setup_build_environment_common(self, env):
         env.unset('LC_ALL')
+        # MPI flags should be the same as run environment, i.e. no wrappers
+        if 'mpi' in self.spec:
+            self.spec['mpi'].package.setup_run_environment(env)
```

Comment:
I am confused about this - if you need something at build time from the package, it should be already set? can you give example where this happens? your osx with brew installed openmpi?

Reply:
Sorry, this fix is worth some explanation. We don't want wrappers with `nrnivmodl`. So we use this to override build-time env vars with runtime ones, so that MPI_CC env (e.g. `OMPI_CC` vars point to the final compiler, not the wrapper.

Reply:
> I agree this feels a bit hacky. It comes from the fact that nrnivmodl doesn't save the original OMPI_CC.... vars.

hmm..ok! I would like to understand this better from @iomaganaris because I thought we filter compilers correctly in neuron installation so that nrnivmodl actually save original compilers and `not spack wrappers`.

Reply:
What happens in `neuron` is that we filter the `nrnmech_makefile` to change the `CC/CXX=` to point to the `mpi` compilers (https://github.com/BlueBrain/spack/blob/7363c10eade107c946c47f2cec26780c43bd5af3/var/spack/repos/builtin/packages/neuron/package.py#L407, https://github.com/BlueBrain/spack/blob/7363c10eade107c946c47f2cec26780c43bd5af3/var/spack/repos/builtin/packages/neuron/package.py#L447). Then when the module is loaded we set in the `run_environment` `MPICC_CC` and `MPICXX_CXX` to get the correct underlying compiler (https://github.com/BlueBrain/spack/blob/7363c10eade107c946c47f2cec26780c43bd5af3/var/spack/repos/builtin/packages/neuron/package.py#L470). The issue I see is that we don't set corresponding variables for the `OpenMPI`, so instead of setting the `MPICC/CXX_CC/CXX` maybe se should run the `setup_run_environment` as Fernando does here to cover all the cases. However I have a question about that. Since `mpi` is a dependency of `neuron` shouldn't the `mpi` package which was used to install `neuron` be loaded as well to set the `MPI` related variables?

Reply:
The issue is that Spack doesnt load modules at build time, instead it creates the EnvironmentModifications structure. If we look inside e.g. openmpi we see that  `setup_build_environment()` sets `MPICC_CC` to the compiler wrappers, which I believe we shouldn't change. So we need to override that somehow, to my mind either from nrnivmodl or the spack package like this. The reason I went for this solution was to avoid another patch in Neuron.

Reply:
> we filter the nrnmech_makefile to change the CC/CXX= to point to the mpi compilers
The reason we cant do the same for MPICC_CXX is because we don't know the relevant var name. E.g. for openmpi it's OMPI_CC. Hence the loading of the runtime settings generically

Reply:
I saw than in the `OpenMPI` package that there is the `setup_dependent_build_environment` in https://github.com/BlueBrain/spack/blob/develop/var/spack/repos/builtin/packages/openmpi/package.py#L317 
which should setup the build environment correctly for the dependent packages. The only issue I see is in runtime, when we run `nrnivmodl` without setting `OMPI_CC` now. Was there any other issue during building or runtime?
For the above issue maybe we should add the `self.spec['mpi'].package.setup_run_environment(env)` in https://github.com/BlueBrain/spack/blob/7363c10eade107c946c47f2cec26780c43bd5af3/var/spack/repos/builtin/packages/neuron/package.py#L470 instead of simply setting the `MPICC/CXX_CC/CXX`

Reply:
> `setup_dependent_build_environment` which should setup the build environment correctly for the dependent packages

`setup_dependent_build_environment` does 
```
env.set('OMPI_CC', spack_cc)
env.set('OMPI_CXX', spack_cxx)
```
which are the compiler wrappers and we want to avoid that, right?

Reply:
I `think` I see what's happening : if you look at the git history of openmpi recipe:

```
commit dfcac4fda2bcdbd6e2d1effede023491e9814db3
Author: Pramod S Kumbhar <pramod.s.kumbhar@gmail.com>
Date:   Mon Mar 12 21:17:32 2018 +0100

    Changes incorporated initial private fork : https://github.com/pramodskumbhar/spack.git

diff --git a/var/spack/repos/builtin/packages/openmpi/package.py b/var/spack/repos/builtin/packages/openmpi/package.py
index 747e8639ac..a0563f2521 100644
--- a/var/spack/repos/builtin/packages/openmpi/package.py
+++ b/var/spack/repos/builtin/packages/openmpi/package.py
@@ -293,6 +293,14 @@ class Openmpi(AutotoolsPackage):
         spack_env.set('OMPI_FC', spack_fc)
         spack_env.set('OMPI_F77', spack_f77)

+    def setup_environment(self, spack_env, run_env):
+        ''' often we use external mpich/openmpi
+            packages and by defualt they are using system clang.
+            we need to set compiler here to avoid link errors
+        '''
+        run_env.set('OMPI_CC', self.compiler.cc)
+        run_env.set('OMPI_CXX', self.compiler.cxx)
```

In the old days, `setup_environment` was called at build as well as runtime. 

Then, Matthias merged this with upstream and we got:

```
commit ceb461a16219546bf0c46034563511366e585e53
Author: Matthias Wolf <m+git@sushinara.net>
Date:   Tue Feb 25 10:14:11 2020 +0100

    core: better accomodation for Travis testing. (#718)
    * core: roll back aggressive caching to fix tests
....
-    def setup_dependent_environment(self, spack_env, run_env, dependent_spec):
-        spack_env.set('MPICC',  join_path(self.prefix.bin, 'mpicc'))
-        spack_env.set('MPICXX', join_path(self.prefix.bin, 'mpic++'))
-        spack_env.set('MPIF77', join_path(self.prefix.bin, 'mpif77'))
-        spack_env.set('MPIF90', join_path(self.prefix.bin, 'mpif90'))
+    def setup_dependent_build_environment(self, env, dependent_spec):
+        env.set('MPICC',  join_path(self.prefix.bin, 'mpicc'))
+        env.set('MPICXX', join_path(self.prefix.bin, 'mpic++'))
+        env.set('MPIF77', join_path(self.prefix.bin, 'mpif77'))
+        env.set('MPIF90', join_path(self.prefix.bin, 'mpif90'))

-        spack_env.set('OMPI_CC', spack_cc)
-        spack_env.set('OMPI_CXX', spack_cxx)
-        spack_env.set('OMPI_FC', spack_fc)
-        spack_env.set('OMPI_F77', spack_f77)
+        env.set('OMPI_CC', spack_cc)
+        env.set('OMPI_CXX', spack_cxx)
+        env.set('OMPI_FC', spack_fc)
+        env.set('OMPI_F77', spack_f77)

-    def setup_environment(self, spack_env, run_env):
+    def setup_run_environment(self, env):
         ''' often we use external mpich/openmpi
             packages and by defualt they are using system clang.
             we need to set compiler here to avoid link errors
         '''
-        run_env.set('OMPI_CC', self.compiler.cc)
-        run_env.set('OMPI_CXX', self.compiler.cxx)
+        env.set('OMPI_CC', self.compiler.cc)
+        env.set('OMPI_CXX', self.compiler.cxx)
```

`setup_environment` got renamed to `setup_run_environment` and `OMPI_CC/CXX` is no longer set to original compilers as I intended.

Reply:
@ferdonline : Considering OpenMPI, was this on your Mac? I know this change is simple and may be we can merge as it is (bit ugly to call run_env by dependent packages) or change `setup_dependent_build_environment` to setup real compilers. 

If this on your Mac, I can check quickly.

Reply:
yep this was on my mac. But I would expect that to happen both in mac and linux. The symptoms is that you see a lot of RPATH entries, many of them duplicated.

Reply:
Either way, I think it is somewhat correct the current behavior of OpenMPI for most situations. It's just that for neurodamus we want it "raw", so I think the patch must be there.